catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardTypeImpl.java

```diff
@@ -208,17 +209,38 @@ private void readObject(ObjectInputStream stream) throws IOException, ClassNotFo
 
         int numElements = stream.readInt();
 
-        descriptors = new HashSet<AttributeDescriptor>();
+        this.descriptors = IntStream.range(0, numElements)
+                .boxed()
+                .map(objectGetter(stream))
+                .collect(ATTRIBUTE_COLLECTOR);
+    }
+
+    private Function<Integer, AttributeDescriptor> objectGetter(ObjectInputStream stream) {
+        return (i) -> {
+            try {
+                return (AttributeDescriptor) stream.readObject();
+            } catch (IOException | ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        };
+    }
+
+    /*
+     * Only to be used by the getDescriptors method.
+     */
+    private Set<AttributeDescriptor> cachedDescriptors;
 
-        for (int i = 0; i < numElements; i++) {
-            descriptors.add((AttributeDescriptor) stream.readObject());
+    private Set<AttributeDescriptor> getDescriptors() {
```

Comment:
As often as `getAttributeDescriptors` is called, I think you could simplify by creating an immutable map and set in the constructors instead of doing a lazy creation.

https://github.com/codice/ddf/search?utf8=%E2%9C%93&q=getAttributeDescriptors